src/api/spec/controllers/source/key_info_controller_controller_spec.rb

```diff
@@ -24,7 +23,7 @@
     before do
       Rails.cache.clear
       # NOTE: we're not using VCR here because the backend does not have the obs signer setup by default
-      stub_request(:get, backend_url).and_return(body: keyinfo_response)
+      allow(Project::KeyInfo).to receive(:key_info_for_project).and_return(keyinfo_response)
```

Comment:
I would prefer to leave the specs as they are. Since the one thing that matters is that the code makes a GET request to the backend for the key_info, so lets test only that.

Reply:
Mmmm... no. What I specially don't like at all is that you should know that the app needs to talk in HTTP with the backend. I rather prefer to "hide" that complexity in a API wrapper like I'm constructing. If in the future we change the way we communicate to the backend we won't have to change a test for it.

Reply:
> What I specially don't like at all is that you should know that the app needs to talk in HTTP with the backend. I rather prefer to "hide" that complexity in an API wrapper like I'm constructing.

This is true for the app but not for the specs. In the specs we should mock external services (i.e. the backend) but not mock internal classes.  

> If in the future we change the way we communicate to the backend we won't have to change a test for it.

If your aim is to make it so that the specs don't specifically mention the use of http as the protocol then I would suggest to create a spec helper method like:
```ruby
def stub_backend_request(method, url)
```

Otherwise the issue I have with this is that before this changes specs allowed you to refactor, they did not care about the internal implementation of classes or stuff like that. They just said "the backend will respond to this request in this way, and the controller should display this value". The specs treat the controller like a black box, only mocking the external inputs and only testing the outputs.

After this change, the specs now say "The `Project::KeyInfo` controller will response with the keyinfo_response if `key_info_for_project` method is called on it and the controller should display this value". This is much more specific to the implementation and does not treat the controller like a black box.

Reply:
Ok, I will stub the request then in the spec I get what you mean. The ugly part is to "calculate" or hold the url in the spec... but anyway.